### PR TITLE
Graphite: Interpolate Graphite alerts with tags

### LINKED ISSUE #32977: Alert notification templating not working with Graphite tags
**What happened**:
I created an alert on a graphite metric with a host tag 
image

and I'm getting this notification:
image

**What you expected to happen**:
The ${host] is replaced by the tag value, as explained in the documentation:
https://grafana.com/docs/grafana/latest/alerting/add-notification-template/ 

**Environment**:
- Grafana version: 7.5.3
- Data source type & version: Graphite

### PR DESCRIPTION
**Pull request by ifrost**
**What this PR does / why we need it**:

This is to refactor Graphite backend queries to data frames. It also adds support for parsing tags which allows to use them as variables in templates (#32977). It will conflict with data frame conversion applied in #35798 but the implementation differs a bit so we might need to do some mix & match :)

In this branch:
* Added handling for null/nil values. Some points might be missing when coming from Graphite. In previous implementation with DataPoints they were mapped to 0 instead of `nil`.
* Invalid timestamp results with an error.
* Also added some unit tests
* Added support for tags (returned by Graphite in a separate structure)

**Which issue(s) this PR fixes**:

Fixes #32977
Subtask of #22918 
Conflicts with #35798

**Special notes for your reviewer**:

How to test alerting?

Run dev environment:

```
make devenv sources=graphite
make run
```

Pass random data to Graphite running [ ./graphite-alerting.sh](https://gist.github.com/ifrost/29e5b4041b0362e968b82ff459360065#file-graphite-alerting-sh)

Configure SMTP.  For example you can use  [mailslurper ](https://github.com/mailslurper/mailslurper) with following custom.ini config:

```
[smtp]
enabled = true
host = localhost:2500
user =
password =
cert_file =
key_file =
skip_verify = false
from_address = admin@grafana.localhost
from_name = Grafana
ehlo_identity =
startTLS_policy =
```

Create Alert notification channel:

Screenshot 2021-06-17 at 15 11 28

Create a dashboard with the following query (make sure to use `tag:environment` when selecting the tag name, or use raw query: `(seriesByTag('environment=prod')`):

Screenshot 2021-06-17 at 15 00 46

Create an alert with variables:

Screenshot 2021-06-17 at 15 05 52

Go back to the script that is sending random data and increase the baseline above 60 (using up/down arrows). Metric values should go up triggering an alert:

Screenshot 2021-06-17 at 15 54 08

In mailslurper (http://localhost:8080) you should see:
Screenshot 2021-06-17 at 15 07 18

⚠️ When alert goes back to "ok" stage the interpolation won't work due to #30606

### CODE REVIEW

**Comment by kylebrandt** [comment]
In terms of new Alerting in 8 (Unified Alerting , `ngalert` feature flag) it seems to work with multi-dimensional alerts:

(Disclaimer: do not know what I am doing with graphite)

image

image

image

**Review by dsotirakis — COMMENTED** [review]
LGTM code-wise! Not approving since it'll need someone from the @grafana/observability-squad to take it for a ride I guess! Great work 🎉

**Review by ying-jeanne — APPROVED** [review]
LGTM instead of returning value 0 use nil would be good. The tag for value column is very specific graphite, with the test of kyle, it seems working fine, i would let severability team give it a check.

File: pkg/tsdb/graphite/graphite.go [hunk 1]
```diff
@@ -242,3 +261,22 @@ func epochMStoGraphiteTime(tr plugins.DataTimeRange) (string, string, error) {
...
+
+	timestamp := time.Unix(int64(dataTimePoint[1].Float64), 0).UTC()
+
+	if dataTimePoint[0].Valid {
+		var value = new(float64)
+		*value = dataTimePoint[0].Float64
+		return timestamp, value, nil
+	} else {
+		return timestamp, nil, nil
+	}
```
Comment (Elfo404) [inline pkg/tsdb/graphite/graphite.go:281]: ❤️

**Review by Elfo404 — APPROVED** [review]
Code looks good, also took it for a ride and seems to work as you described, but i'm also not extremely comfortable with graphite

good stuff! 🚀

**Pull request merged**